### DISCUSSION TITLE
How to run in a notebook on a remote supercomputer (no root access)

### DISCUSSION THREAD

**Discussion by rabernat** [pyvista/pyvista#1406] [Q&A] [answered]
I would like to run pyvista within a jupyter notebook on my university's HPC system. The [instructions for running on remote systems](https://docs.pyvista.org/getting-started/installation.html) focus on Docker containers. They assume I have root and can install xvfb. Neither is the typically the case for HPC systems. 

Are there any workarounds for this case? Or is it not possible to use pyvista in this situation?

**Comment by akaszynski** [comment]
This is a perfect use case for using `ipygany`.  See the docs here:
https://dev.pyvista.org/jupyter/ipygany.html

This is currently implemented on our release candidate, with a stable release coming (probably) next week.  Install with:
```
pip install pyvista==0.30.0rc1
```

`ipygany` support is somewhat limited right now, so you won't see all features implemented (like lighting, shadows, etc.)  We're working to provide better support, and if you can provide us with some feedback that would be really helpful.

**Comment by rabernat** [comment]
> This is a perfect use case for using `ipygany`

Thanks for the tip. I'll try it.

> This is currently implemented on our release candidate

Can you clarify what you mean by "this"?

While ipygany sounds cool, I'm curious if there any way to use server-side rendering in a notebook without root access.

**Comment by akaszynski** [comment]
> Can you clarify what you mean by "this"?

The `ipygany` renderer, which uses client side rendering with three.js

> While ipygany sounds cool, I'm curious if there any way to use server-side rendering in a notebook without root access.

Try this to install xvfb without root:
https://serverfault.com/questions/828990/how-to-install-xvfb-without-root-access-tried-with-linuxbrew

**Comment by hakostra** [comment]
Being in the same situation myself, I have built VTK with OSMesa software rendering support. This VTK build then runs "out of the box" without any graphics hardware, X libraries or xvfb at all. Since it also can be useful for others, I uploaded my build scripts here:
https://github.com/hakostra/vtk_build
I also uploaded a pre-built wheel for Python 3.8. Feel free to use it (but no warranties given).

**Comment by LTMeyer** [comment]
Hi everyone,

I would like to use pyvista to generate screenshot from vtu files using a python script on cluster nodes. I am not using jupyter, and I cannot install software mentioned in the documentation (namely `libgl1-mesa-dev` and `xvfb`). Is there still any possibility to use pyvista to generate screenshots on a remote server without graphic display functionalities?

**Comment by akaszynski** [comment]
Install the EGL vtk wheels available in one of our old releases and you will be able to plot `off_screen` without having a virtual framebuffer setup.
https://github.com/pyvista/pyvista/releases/tag/0.27.0

**Comment by LTMeyer** [comment]
> Install the EGL vtk wheels available in one of our old releases and you will be able to plot `off_screen` without having a virtual framebuffer setup.
> https://github.com/pyvista/pyvista/releases/tag/0.27.0

Thank you very much for your quick answer. 
Indeed after having installed the wheel, I was able to instantiate the `Plotter` with the `off_screen` option without any warning nor segfault. Unfortunately, once I tried to plot and save a screenshot, still off screen, I got a segfault error : 
> ERROR:root:Only EGL 1.4 and greater allows OpenGL as client API. See eglBindAPI for more information.
> ERROR:root:GLEW could not be initialized: Missing GL version
It seems graphic libraries are still required. Unfortunately I cannot install them on the cluster.

It may be naive, but is there any way to just retrieve the image that would be plot as an array and save it with `PIL` or other library? I am trying to plot a mesh with P2 Lagrangian elements (i.e. triangles with 6 points). This is not possible with `matplotlib` which only allow to draw mesh with regular triangles (i.e. 3 points per triangle). Hence, I considered using `pyvista` for the plotting.

**Comment by hakostra** [comment]
Did you see my prev. post/comment? I have created a VTK wheel with bundled OSMesa for this specific purpose. This does all rendering on the CPU and does not require any X libraries or GPU hardware. See https://github.com/hakostra/vtk_build

**Reply by akaszynski** [reply]
Awesome! Thanks for adding that.

**Comment by josephko91** [comment]
Just wondering, is there an updated workaround for this? (That doesn't require the custom VTK wheel shared by @hakostra ). I faced similar issues with visualization in an HPC environment without root privileges.

**Comment by hakostra** [comment, marked as answer]
Kitware now publishes a custom VTK package built with OSMesa:
https://discourse.vtk.org/t/status-update-vtk-python-wheels/11212

So I consider my own package obsolete. I have switched to using the "official" one.

**Reply by MatthewFlamm** [reply]
When using pyvista, it is important to note the dance of uninstalling vtk before installing vtk-osmesa or using `--no-deps` and explicitly installing the dependencies.

```
pip install pyvista
pip uninstall vtk -y
pip install --extra-index-url https://wheels.vtk.org vtk-osmesa
```

Added this so this discussion chain would have a copy-pastable code, and users can visit the link provided for more details.

**Reply by jourdain** [reply]
Side add-on, we also have a `vtk-egl` wheel too.
